### PR TITLE
Don't delegate visitCollection -> visitGenericFileTree

### PR DESCRIPTION
**Pull request by wolfs**
in FilteredMinimalFileTree.

This came up in https://github.com/gradle/gradle/issues/19158 when I tried to remove `visitGenericFileTree()`. Actually, it seems like it is impossible to get to the call of the method, since
- Only `GeneratedSingletonFileTree` calls `visitCollection()` when `prepareForVisit` returns `FileCollectionStructureVisitor.VisitType.NoContents`.
- Only [`CollectingVisitor`](https://github.com/gradle/gradle/blob/b7e1a80ddd35b0c17177a103d9ce63e11fa6ce48/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/FileCollectionCodec.kt#L163) uses `NoContents`, though that already returns early since `startVisit` there will end up in the branch for [`FileTreeInternal`](https://github.com/gradle/gradle/blob/b7e1a80ddd35b0c17177a103d9ce63e11fa6ce48/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/FileCollectionCodec.kt#L143-L146), aborting the visiting.

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by wolfs** [comment]
@bot-gradle test RfM

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/FilteredMinimalFileTree.java [hunk 1]
```diff
@@ -74,7 +74,7 @@ public VisitType prepareForVisit(FileCollectionInternal.Source source) {
 
             @Override
             public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
```
Comment (lptr) [inline subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/FilteredMinimalFileTree.java:76]: Wouldn't it make sense to introduce a `MinimalFileTreeStructureVisitor` here that doesn't even have this method?

**Review by lptr — COMMENTED** [review]
The only place we create a `GeneratedSingletonFileTree` is here:

https://github.com/gradle/gradle/blob/457f85663d4521d4876924689bfcc1effc4ee53a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/DefaultFileCollectionFactory.java#L213-L216

...where we immediately wrap it in a `FileTreeAdapter`. Can we refactor `GeneratedSingletonFileTree` to extend `AbstractFileTree` directly?

**Comment by wolfs** [comment]
> ...where we immediately wrap it in a FileTreeAdapter. Can we refactor GeneratedSingletonFileTree to extend AbstractFileTree directly?

We could, though I don't see the point of doing so. The point of having something like `MinimalFileTree` and `FileTreeAdapter` is to simplify creating different types of file trees easily. I don't see how implementing FileTreeInternal in GeneratedSingletonFileTree directly would make things easier.

File: subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/FilteredMinimalFileTree.java (hunk 1 above)
Comment (wolfs) [reply subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/FilteredMinimalFileTree.java]: Good idea, I introduced this interface.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Pull request merged**